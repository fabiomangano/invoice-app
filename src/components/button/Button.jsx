/* eslint-disable react/button-has-type */
/* https://github.com/yannickcr/eslint-plugin-react/issues/1846 */
import React from 'react';
import PropTypes from 'prop-types';
import cn from 'classnames';
import './Button.style.scss';

const Button = ({ children, type, onClick, variant, size }) => (
  <button
    className={cn(
      'custom-btn',
      `custom-btn--${variant}`,
      `custom-btn--${size}`,
    )}
    type={type}
    onClick={onClick}
  >
    {children}
  </button>
);

Button.defaultProps = {
  type: 'button',
  variant: 'primary',
  size: 'medium',
};

Button.propTypes = {
  type: PropTypes.oneOf(['button', 'reset', 'submit']),
  variant: PropTypes.oneOf(['primary', 'secondary', 'tertiary', 'danger']),
  size: PropTypes.oneOf(['small', 'medium', 'large', 'fluid']),
  children: PropTypes.node.isRequired,
  onClick: PropTypes.func.isRequired,
};

export default Button;
