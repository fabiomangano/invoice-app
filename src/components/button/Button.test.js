import React from 'react';
import Renderer from 'react-test-renderer';
import { mount } from 'enzyme';
import Button from './Button';

describe('Custom Button', () => {
  let tree;
  let onClick;

  beforeAll(() => {
    onClick = jest.fn();
  });

  afterEach(() => {
    tree = null;
  });

  it('should render a custom button with default props', () => {
    Renderer.act(() => {
      tree = Renderer.create(<Button onClick={onClick}>Default Button</Button>);
    });

    expect(tree.toJSON()).toMatchSnapshot();
  });

  it('should render a primary button', () => {
    Renderer.act(() => {
      tree = Renderer.create(
        <Button onClick={onClick} variant="primary">
          Primary Button
        </Button>,
      );
    });

    expect(tree.toJSON()).toMatchSnapshot();
  });

  it('should render a secondary button', () => {
    Renderer.act(() => {
      tree = Renderer.create(
        <Button onClick={onClick} variant="secondary">
          Secondary Button
        </Button>,
      );
    });

    expect(tree.toJSON()).toMatchSnapshot();
  });

  it('should render a tertiary button', () => {
    Renderer.act(() => {
      tree = Renderer.create(
        <Button onClick={onClick} variant="tertiary">
          Tertiary Button
        </Button>,
      );
    });

    expect(tree.toJSON()).toMatchSnapshot();
  });

  it('should render a danger button', () => {
    Renderer.act(() => {
      tree = Renderer.create(
        <Button onClick={onClick} variant="danger">
          Danger Button
        </Button>,
      );
    });

    expect(tree.toJSON()).toMatchSnapshot();
  });

  it('should render a small button', () => {
    Renderer.act(() => {
      tree = Renderer.create(
        <Button onClick={onClick} size="small">
          Small Button
        </Button>,
      );
    });

    expect(tree.toJSON()).toMatchSnapshot();
  });

  it('should render a medium button', () => {
    Renderer.act(() => {
      tree = Renderer.create(
        <Button onClick={onClick} size="medium">
          Medium Button
        </Button>,
      );
    });

    expect(tree.toJSON()).toMatchSnapshot();
  });

  it('should render a large button', () => {
    Renderer.act(() => {
      tree = Renderer.create(
        <Button onClick={onClick} size="large">
          Large Button
        </Button>,
      );
    });

    expect(tree.toJSON()).toMatchSnapshot();
  });

  it('should render a fluid button', () => {
    Renderer.act(() => {
      tree = Renderer.create(
        <Button onClick={onClick} size="fluid">
          Fluid Button
        </Button>,
      );
    });

    expect(tree.toJSON()).toMatchSnapshot();
  });

  it('should trigger cb function onClick', () => {
    const wrapper = mount(<Button onClick={onClick}>Click</Button>);
    wrapper.find('button').simulate('click');
    expect(onClick).toHaveBeenCalledTimes(1);
  });
});
