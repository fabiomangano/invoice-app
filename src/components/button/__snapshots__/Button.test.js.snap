// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Custom Button should render a custom button with default props 1`] = `
<button
  className="custom-btn custom-btn--primary custom-btn--medium"
  onClick={[MockFunction]}
  type="button"
>
  Default Button
</button>
`;

exports[`Custom Button should render a danger button 1`] = `
<button
  className="custom-btn custom-btn--danger custom-btn--medium"
  onClick={[MockFunction]}
  type="button"
>
  Danger Button
</button>
`;

exports[`Custom Button should render a fluid button 1`] = `
<button
  className="custom-btn custom-btn--primary custom-btn--fluid"
  onClick={[MockFunction]}
  type="button"
>
  Fluid Button
</button>
`;

exports[`Custom Button should render a large button 1`] = `
<button
  className="custom-btn custom-btn--primary custom-btn--large"
  onClick={[MockFunction]}
  type="button"
>
  Large Button
</button>
`;

exports[`Custom Button should render a medium button 1`] = `
<button
  className="custom-btn custom-btn--primary custom-btn--medium"
  onClick={[MockFunction]}
  type="button"
>
  Medium Button
</button>
`;

exports[`Custom Button should render a primary button 1`] = `
<button
  className="custom-btn custom-btn--primary custom-btn--medium"
  onClick={[MockFunction]}
  type="button"
>
  Primary Button
</button>
`;

exports[`Custom Button should render a secondary button 1`] = `
<button
  className="custom-btn custom-btn--secondary custom-btn--medium"
  onClick={[MockFunction]}
  type="button"
>
  Secondary Button
</button>
`;

exports[`Custom Button should render a small button 1`] = `
<button
  className="custom-btn custom-btn--primary custom-btn--small"
  onClick={[MockFunction]}
  type="button"
>
  Small Button
</button>
`;

exports[`Custom Button should render a tertiary button 1`] = `
<button
  className="custom-btn custom-btn--tertiary custom-btn--medium"
  onClick={[MockFunction]}
  type="button"
>
  Tertiary Button
</button>
`;
